### PR TITLE
Improve table JSON name mangling

### PR DESCRIPTION
**Pull request by AdrianVovk**
See commit messages

### CODE REVIEW

File: src/import/importctl.c [hunk 1]
```diff
@@ -908,6 +908,11 @@ static int list_images(int argc, char *argv[], void *userdata) {
         (void) table_set_sort(t, (size_t) 0, (size_t) 1);
         table_set_ersatz_string(t, TABLE_ERSATZ_DASH);
 
+        /* Starting in v257, these fields would be automatically formatted with underscores. However, this
+         * command was introduced in v256, so changing the field name would be a breaking change. */
+        (void) table_set_json_field_name(t, 8, "usage-exclusive");
+        (void) table_set_json_field_name(t, 10, "limit-exclusive");
```
Comment (YHNdnzj) [inline src/import/importctl.c:918]: I don't feel the necessity to do this - we don't promise the same level of compatibility as public APIs for CLI tools' JSON output (e.g. https://github.com/systemd/systemd/pull/30958), and if some json parsers don't like dashes we'd better translate them here too.

File: src/shared/format-table.c [hunk 2]
```diff
@@ -2883,16 +2889,30 @@ static int table_data_to_json(TableData *d, sd_json_variant **ret) {
...
+         *  - Convert spaces to underscores
+         *  - Convert dashes to underscores (some JSON parsers don't like dealing with dashes)
+         *  - Make the first letter of each words lowercase
+         */
+
+        bool new_word = true;
+        char *c;
 
-        char *c = strdup(f);
+        c = strdup(f);
```
Comment (YHNdnzj) [inline src/shared/format-table.c]: Not your fault, but assert(f)

File: src/shared/format-table.c [hunk 3]
```diff
@@ -2883,16 +2889,30 @@ static int table_data_to_json(TableData *d, sd_json_variant **ret) {
...
+        for (char *x = c; *x; x++) {
+                if (new_word) {
+                        *x = ascii_tolower(*x);
+                        new_word = false;
+                }
+
+                if (isspace(*x) || *x == '-') {
                         *x = '_';
+                        new_word = true;
+                }
```
Comment (YHNdnzj) [inline src/shared/format-table.c]: I'd place this before `new_word` branch, and continue on space/dash

File: src/shared/format-table.c [hunk 4]
```diff
@@ -2889,16 +2889,33 @@ static int table_data_to_json(TableData *d, sd_json_variant **ret) {
 
 static char* string_to_json_field_name(const char *f) {
         /* Tries to make a string more suitable as JSON field name. There are no strict rules defined what a
-         * field name can be hence this is a bit vague and black magic. Right now we only convert spaces to
-         * underscores and leave everything as is. */
+         * field name can be hence this is a bit vague and black magic. Here's what we do:
+         *  - Convert spaces to underscores
+         *  - Convert dashes to underscores (some JSON parsers don't like dealing with dashes)
```
Comment (poettering) [inline src/shared/format-table.c:2902]: hmm, did you verify that this doesn't break any of our existing tables? we should really stick to the current json field names for existing tables. It definitely makes sense to default to the rules that you are proposing for new tables, but for existing tables we should continue to generate the same (flawed) field names as we always did.

File: src/shared/format-table.c [hunk 5]
```diff
@@ -2889,16 +2889,33 @@ static int table_data_to_json(TableData *d, sd_json_variant **ret) {
 
 static char* string_to_json_field_name(const char *f) {
```
Comment (poettering) [inline src/shared/format-table.c]: maybe export this and add a quick test case that formats a bunch of names

File: src/shared/format-table.c [hunk 6]
```diff
@@ -2889,16 +2889,33 @@ static int table_data_to_json(TableData *d, sd_json_variant **ret) {
...
+        assert(f);
+
+        c = strdup(f);
         if (!c)
                 return NULL;
 
-        for (char *x = c; *x; x++)
-                if (isspace(*x))
+        for (char *x = c; *x; x++) {
+                if (isspace(*x) || *x == '-') {
```
Comment (poettering) [inline src/shared/format-table.c]: you probably want to trigger this on `_` too, no? so that new_word is set true?

**Comment by poettering** [comment]
i like where this is going, just a bit afraid that existing tables might be affected by this.

a quick grep for `table_new` reveals to me that some existing tables might be affected by this:

1. in src/analyze/analyze-image-policy.c there's a table with a "read-only" column
2. src/import/importctl.c has "usage-exclusive"
3. src/mount/mount-tool.c has a table will uppercase columns
4. src/sysupdate/sysupdate.c has a table tries-done/tries-left

i guess this could be quickly

File: src/shared/format-table.c (hunk 4 above)
Comment (YHNdnzj) [reply src/shared/format-table.c:2902]: Ugh, I requested to always use the new format in https://github.com/systemd/systemd/pull/33609#discussion_r1664883795

Comment (poettering) [reply src/shared/format-table.c:2902]: ah, didn't see that.

File: src/import/importctl.c (hunk 1 above)
Comment (poettering) [reply src/import/importctl.c:918]: well, our json output only really exists to be consumable/parsable by other tools. To make this workable we really should try to make it stable (unless it's so broken that it was unusable before)

Comment (AdrianVovk) [reply src/import/importctl.c:918]: so should I put this back?

**Comment by AdrianVovk** [comment]
Yeah I did the same grepping (also grepped for `TABLE_HEADING` and `TABLE_FIELD`), and ended up with roughly the same list. Then I checked each of those.

`analyze-image-policy.c`: Never printed as JSON - only as a table. So no break.

`importctl.c`: I added a work-around for it, but was told to remove it. Let me know if I should add it back.

`mount-tool.c` : Never printed as JSON. So no break.

`sysupdate.c`: honestly thought the JSON output for sysupdate came as part of #33520, but looks like this piece bit was there previously...

In both `importctl` and `sysupdate` cases, I'm pretty confident that very few people - if any - are using these JSON fields. But I guess if we consider them API-enough to warrant the workaround then I can add the workaround.

File: src/import/importctl.c (hunk 1 above)
Comment (poettering) [reply src/import/importctl.c:918]: I think that would be best, yes.

File: src/shared/format-table.h [hunk 7]
```diff
@@ -74,6 +74,7 @@ typedef enum TableDataType {
         TABLE_SET_BOTH_UNDERLINES,
         TABLE_SET_URL,
         TABLE_SET_UPPERCASE,
+        TABLE_SET_JSON_FIELD_NAME,
```
Comment (keszybz) [inline src/shared/format-table.h:77]: This should come with a unittest somewhere.

File: src/import/importctl.c (hunk 1 above)
Comment (keszybz) [reply src/import/importctl.c:918]: FWIW, I think it's fine if we change the naming. Not in a stable release, but in a new upstream release. I think the number of people who use this and are not upstream developers is miniscule. If  you decide to keep the override, I'm fine with that too ;)

**Review by keszybz — CHANGES_REQUESTED** [review]
I think the only firm conclusion it that we need a bunch of unitttests for this.

**Comment by AdrianVovk** [comment]
Alright. Improved the heuristic, added test cases, looked again for places where the heuristic would produce different output than before and re-introduced a commit fixing those places

**Comment by poettering** [comment]
CI very unhappy with this.

**Comment by poettering** [comment]
other than the ci issue: lgtm (apparently i am not as good as a CI, I see nothing wrong here ;-))

**Comment by AdrianVovk** [comment]
CI Failures appear unrelated:

`mkosi / ci (fedora, 40, 0, -Og, yes)`:
`92/92 systemd:integration-tests / TEST-84-STORAGETM     FAIL`

`mkosi / ci (fedora, rawhide, address,undefined, 1, -Og, yes)`:
`14/92 systemd:integration-tests / TEST-85-NETWORK-NetworkdMatchTests     FAIL`

`noble-arm64`:
`13799s networkd-test.py     FAIL non-zero exit status 1`

File: src/shared/format-table.c [hunk 8]
```diff
@@ -1184,6 +1184,17 @@ int table_add_many_internal(Table *t, TableDataType first_type, ...) {
                         goto check;
                 }
 
+                case TABLE_SET_JSON_FIELD_NAME: {
+                        const char *n = va_arg(ap, const char*);
+                        size_t idx;
+                        if (t->vertical)
+                                idx = TABLE_CELL_TO_INDEX(last_cell) / t->n_columns - 1;
```
Comment (poettering) [inline src/shared/format-table.c:1192]: hmm, please add an assert() here that gates the underflow (i.e. if TABLE_CELL_TO_INDEX(last_cell) is <  t->n_columns)

i.e. it's fine to not handle the underflow, after all that is a programmer bug. but please catch it with an assert here.

File: src/test/test-format-table.c [hunk 9]
```diff
@@ -404,6 +404,38 @@ TEST(json) {
         assert_se(sd_json_variant_equal(v, w));
 }
 
+TEST(json_mangling) {
+        struct {
```
Comment (poettering) [inline src/test/test-format-table.c]: static const

File: src/test/test-format-table.c [hunk 10]
```diff
@@ -404,6 +404,38 @@ TEST(json) {
...
+                { "Foo (Bar)", "foo__bar_"},
+                { "MixedCase ALLCAPS", "mixedCase_ALLCAPS" },
+                { "_X", "_x" },
+                { "_Foo", "_foo" },
+        };
+
+        FOREACH_ELEMENT(i, cases) {
+                _cleanup_free_ char *ret = NULL;
+                assert_se(ret = table_mangle_to_json_field_name(i->arg));
+                printf("\"%s\" -> \"%s\"", i->arg, ret);
```
Comment (poettering) [inline src/test/test-format-table.c]: \n

**Comment by poettering** [comment]
lgtm, just some nitpicks

File: src/shared/format-table.c (hunk 8 above)
Comment (AdrianVovk) [reply src/shared/format-table.c:1192]: I also added another check, in the !vertical case, to make sure TABLE_SET_JSON_FIELD_NAME is only applied onto a header cell

**Comment by AdrianVovk** [comment]
Also rebased on main, to run a more up-to-date CI

**Pull request merged**